### ISSUE TITLE
Consistency of QNode return value shape after qml.gradients.param_shift_hessian() for QNodes that return a one element list

### ISSUE THREAD

**Issue by cvjjm** [PennyLaneAI/pennylane#2194]
### Feature details

For the special case of QNodes returning a list with a single expval the shape returned after transforming such QNode with `qml.gradients.param_shift_hessian()` is a bit "inconsistent" with how the untransformed return value is treated.

What I mean is best explained with the example below.

I think it would be nice if this minor "inconsistency" in the user interface could be fixed.

### How important would you say this feature is?

1: Not important. Would be nice to have.

### Additional information

For me the expected behavior would have been that all asserts pass in the following code: 
```python
import pennylane as qml
from pennylane import numpy as np

wires = 3
dev = qml.device('default.qubit', wires=wires)

def ansatz(params):
    for wire in range(wires):
        qml.Rot(*params[0][wire], wires=wire)
    qml.CNOT(wires=[0,1])
    for wire in range(wires):
        qml.Rot(*params[1][wire], wires=wire)
    qml.CNOT(wires=[1,2])

@qml.qnode(dev)
def cost1(params):
    ansatz(params)
    return qml.expval(qml.PauliZ(0))

@qml.qnode(dev)
def cost2(params):
    ansatz(params)
    return [qml.expval(qml.PauliZ(0))]

@qml.qnode(dev)
def cost3(params):
    ansatz(params)
    return [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))]


params = np.random.rand(2, wires, 3)

expval1 = cost1(params)
expval2 = cost2(params)
expval3 = cost3(params)

assert expval1 == expval2[0]
assert expval1 == expval3[0]

hessian1 = qml.gradients.param_shift_hessian(cost1)(params)
hessian2 = qml.gradients.param_shift_hessian(cost2)(params)
hessian3 = qml.gradients.param_shift_hessian(cost3)(params)

print("hessian1.shape", hessian1.shape, "hessian2.shape", hessian2.shape, "hessian3.shape", hessian3.shape)
assert hessian1.shape == hessian3[0].shape
assert hessian1.shape == hessian2[0].shape, "For consistency with how a QNode with 'return [qml.expval(qml.PauliZ(0))] behaves when it is evaluated (it reutrns a one element tensor) it would be nice that if after transforming it with gradients.param_shift_hessian() it should also return a tensor, whose only element is the hessian. Instead it returns the plain hessian directly.'"
assert hessian1 == hessian2[0]
```
However `hessian1.shape == hessian3.shape != hessian2[0].shape`.

**Comment by dime10** [comment]
Hi @cvjjm, thank you very much for pointing out this inconsistency!

After investigating I've realised the problem is actually not specific to `param_shift_hessian`, `param_shift` for example runs into same issue (but the two are also inconsistent w.r.t. each other in how they behave). Essentially, `param_shift_hessian` will squeeze any dimensions of size 1 from the result, while `param_shift` squeezes in some circumstances but not in others, as shown in the table below:

| QNode return statement  | QNode output shape | param_shift output shape | param_shift_hessian output shape | execution result shape |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| scalar (e.g. `expval`)  | (0,) | __(1, *) !__ | (*, *) | __(#, 1) !__ |
| single element (scalar) list/tuple (e.g. `[expval]`) | (1,) | (1, *) | __(*, *) !__ | (#, 1) |
| multi element (scalar) list/tuple (e.g. `[expval, expval]`) | (2,) | (2, *) | (2, *, *) | (#, 2) |
| non-scalar (e.g. `probs(0)`) | (2,) | (2, *) | (2, *, *) | __(#, 1, 2) !__ |
| single element (non-scalar) list/tuple (e.g. `[probs(0)]`) | (1, 2) | __(2, *) !__ | __(2, *, *) !__ | (#, 1, 2) |
| multi element (non-scalar) list/tuple (e.g. `[probs(0), probs(1)]`) | (2, 2) | (2, 2, *) | (2, 2, *, *) | (#, 2, 2) |
| __bold ! values deviate from expected__| | \* = QNode args shape | \* = QNode args shape | # = number of tapes |

Unfortunately, I don't think either transform is able to detect whether, for example, a single expectation value was wrapped in a list or not. It appears that the QNode output shape that the transforms receive is determined by the batched tape execution (`batch_transform` or maybe `qml.execute`?). @josh146 do you know if it would be possible to make changes such that the tape execution results produce the same QNode output dimensions as when running the QNode directly?

**Comment by josh146** [comment]
wow nice analysis @dime10!

In the case `gradient_transform(qnode)(*args)`, I _think_ this is something we could address.

In `qml.gradients.gradient_transform`, we define the logic for the gradient transform when it acts directly on a QNode:

https://github.com/PennyLaneAI/pennylane/blob/c6d5cc4dfc1dbe15050c53301d8c3b245a8674ef/pennylane/gradients/gradient_transform.py#L113-L116

At the very bottom, we have the potential to modify the return shape:

https://github.com/PennyLaneAI/pennylane/blob/c6d5cc4dfc1dbe15050c53301d8c3b245a8674ef/pennylane/gradients/gradient_transform.py#L158-L160

(tagging @dwierichs here since he might be able to provide more context behind `safe_squeeze`!). The same applies for [`hessian_transform`](https://github.com/PennyLaneAI/pennylane/blob/c6d5cc4dfc1dbe15050c53301d8c3b245a8674ef/pennylane/gradients/hessian_transform.py).

What we could do:

- The QNode, when it executes the qfunc, stores the output as [`qnode._qfunc_output`](https://github.com/PennyLaneAI/pennylane/blob/c6d5cc4dfc1dbe15050c53301d8c3b245a8674ef/pennylane/qnode.py#L488).

- Inside the `default_qnode_wrapper` in both `gradient_transform` and `hessian_transform`, we could inspect this attribute, and use it to adjust the output of the transform:

  ```python
  if isinstance(qnode._qfunc_output, Sequence):
      return_length = len(qnode._qfunc_output)
  ```

**Comment by dwierichs** [comment]
Hi @dime10,
_(this comment only looks at `param_shift`)_
I wanted to look into this and found that `param_shift` produces an output with shape `(*)`, i.e. the expected one, for the following example
```python
dev = qml.device('default.qubit', wires=2)

@qml.qnode(dev)
def circuit(x):
    qml.RX(x[0], wires=0)
    return qml.expval(qml.PauliZ(0))

x = pnp.array([0.3, 0.2, 0.1], requires_grad=True)
```
```pycon
>>> print(qml.gradients.param_shift(circuit)(x))
[-0.29552021  0.          0.        ]
>>> print(qml.gradients.param_shift(circuit)(x).shape)
(3,)
```
When wrapping the `expval` in a list, I actually get the same as the above, so the field in the table one row below would be tagged as unexpected. When adding a second expectation value to the qnode output, I obtain 2D derivatives as expected.
For `probs` I observe the same behaviour as you describe in the table, which is expected according to the rationale of unwrapping outputs with leading dimension 1, as far as I can tell.
Maybe I'm getting the use case wrong?

**Comment by dime10** [comment]
Thanks @josh146 for the tip regarding `[qnode._qfunc_output](https://github.com/PennyLaneAI/pennylane/blob/c6d5cc4dfc1dbe15050c53301d8c3b245a8674ef/pennylane/qnode.py#L488)`, which should help us resolve this issue :)

Oh thanks for the additional datapoint @dwierichs. This seems to be another inconsistency in the output of `param_shift`. For the table I was using the following example:
```py
dev = qml.device('default.qubit', wires=2)
@qml.qnode(dev)
def cost(params):
    qml.Rot(*params, wires=0)
    return ...
params = np.random.rand(3)
qml.gradients.param_shift(cost)(params).shape
```
Using `Rot` instead of `RX` changes the output to `(1,3)` instead of the `(3,)` that your example produces.

**Comment by cvjjm** [comment]
I would be in favor of not unpacking outputs of QNode that return a list with a single item as in `return [qml.expval(...)]`.

The rational and use case is that of QNodes that are dynamically created by some higher level code so that the length of the list or returned expvals is not known in advance and whose output is to be processed by further code. That further code needs to get the QNode output in a consistent shape irrespective of whether the list container 4, twelve, or a single expval. It would be nice if the first dimension would always be the length of the list after `return`.
